src/proof_rules/mod.rs

```diff
@@ -90,33 +90,125 @@ pub trait Encoding: fmt::Debug {
     fn is_terminator(&self) -> bool;
 }
 
+#[derive(Debug, Clone)]
+pub enum EncodingKind {
+    Invariant(InvariantAnnotation),
+    KInduction(KIndAnnotation),
+    Unroll(UnrollAnnotation),
+    Omega(OmegaInvAnnotation),
+    Ost(OSTAnnotation),
+    Past(PASTAnnotation),
+    Ast(ASTAnnotation),
+}
+
+impl EncodingKind {
```

Comment:
I feel these `impl`s are pretty verbose we'll need to think of a better approach to architect the types here. A very simple alternative would be to simply add a method to the `Encoding` trait called `check_calculus` that receives a simple value of a new enum type `Calculus` and a direction and then returns `Ok(())` or an `Err(EncodingCalculusError)`. Then the checks wouldn't be part of a specific calculus, but rather of the encoding.